scene/main/viewport.cpp

```diff
@@ -2607,39 +2595,73 @@ void Viewport::_drop_mouse_focus() {
 void Viewport::_drop_physics_mouseover(bool p_paused_only) {
 	physics_has_last_mousepos = false;
 
+	_cleanup_mouseover_colliders(true, p_paused_only);
+
+#ifndef _3D_DISABLED
+	if (physics_object_over.is_valid()) {
+		CollisionObject3D *co = Object::cast_to<CollisionObject3D>(ObjectDB::get_instance(physics_object_over));
+		if (co) {
+			if (!(p_paused_only && co->can_process())) {
+				co->_mouse_exit();
+				physics_object_over = ObjectID();
+				physics_object_capture = ObjectID();
+			}
+		}
+	}
+#endif
+}
+
+void Viewport::_cleanup_mouseover_colliders(bool clean_all_frames, bool p_paused_only, uint64_t frame_reference) {
 	List<Map<ObjectID, uint64_t>::Element *> to_erase;
 
 	for (Map<ObjectID, uint64_t>::Element *E = physics_2d_mouseover.front(); E; E = E->next()) {
+		if (!clean_all_frames && E->get() == frame_reference) {
+			continue;
+		}
+
 		Object *o = ObjectDB::get_instance(E->key());
 		if (o) {
 			CollisionObject2D *co = Object::cast_to<CollisionObject2D>(o);
 			if (co) {
-				if (p_paused_only && co->can_process()) {
+				if (clean_all_frames && p_paused_only && co->can_process()) {
 					continue;
 				}
 				co->_mouse_exit();
-				to_erase.push_back(E);
 			}
 		}
+		to_erase.push_back(E);
 	}
 
 	while (to_erase.size()) {
 		physics_2d_mouseover.erase(to_erase.front()->get());
 		to_erase.pop_front();
 	}
 
-#ifndef _3D_DISABLED
-	if (physics_object_over.is_valid()) {
-		CollisionObject3D *co = Object::cast_to<CollisionObject3D>(ObjectDB::get_instance(physics_object_over));
-		if (co) {
-			if (!(p_paused_only && co->can_process())) {
-				co->_mouse_exit();
-				physics_object_over = ObjectID();
-				physics_object_capture = ObjectID();
+	// Per-shape
+	List<Map<Pair<ObjectID, int>, uint64_t, PairSort<ObjectID, int>>::Element *> shapes_to_erase;
+
+	for (Map<Pair<ObjectID, int>, uint64_t, PairSort<ObjectID, int>>::Element *E = physics_2d_shape_mouseover.front(); E; E = E->next()) {
+		if (!clean_all_frames && E->get() == frame_reference) {
+			continue;
+		}
+
+		Object *o = ObjectDB::get_instance(E->key().first);
+		if (o) {
+			CollisionObject2D *co = Object::cast_to<CollisionObject2D>(o);
+			if (co) {
+				if (p_paused_only && co->can_process()) {
```

Comment:
Should this check be `if (clean_all_frames && p_paused_only && co->can_process())` in order to follow the same logic as for the mouse_exit events?

Reply:
Yeah, thanks, let's make them consistent, done.

Note that on the logic level these two have the same behavior as `p_paused_only => clean_all_frames` (I only pass p_paused_only when clean_all_frames is true).